### PR TITLE
On publish failures, log error and count them as sys exceptions

### PR DESCRIPTION
**Pull request by srkukarni**
### Motivation
When a user explicit publish fails, we fail in silence currently. This pr fixes that by logging it as well as treating it as a system failure. In python, we already do the same, so this brings both languages to the same behaviour.

*Explain here the context, and why you're making that change. What is the problem you're trying to solve.*

### Modifications

*Describe the modifications you've done.*

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API: (yes / no)
  - The schema: (yes / no / don't know)
  - The default values of configurations: (yes / no)
  - The wire protocol: (yes / no)
  - The rest endpoints: (yes / no)
  - The admin cli options: (yes / no)
  - Anything that affects deployment: (yes / no / don't know)

### Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)
  - If a feature is not applicable for documentation, explain why?
  - If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

### CODE REVIEW

**Comment by srkukarni** [comment]
run integration tests

**Comment by srkukarni** [comment]
run integration tests

File: pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/ContextImpl.java [hunk 1]
```diff
@@ -359,7 +362,11 @@ public ByteBuffer getState(String key) {
             }
         }
 
-        return producer.sendAsync(object).thenApply(msgId -> null);
+        return producer.sendAsync(object).exceptionally(e -> {
```
Comment (merlimat) [inline pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/ContextImpl.java]: This is returning a future that is masking the eventual error. It should be something like : 

```java
CompletableFuture<Void> future = producer.sendAsync(object).thenApply(msgId -> null);
future.exceptionally(e -> {
     this.statsManager.incrSysExceptions(e);
     logger.error("Failed to publish to topic {} with error {}", topicName, e);
     return null;
});
return future;
```

**Comment by srkukarni** [comment]
run java8 tests
run integration tests

**Comment by srkukarni** [comment]
run java8 tests

**Comment by srkukarni** [comment]
run java8 tests

**Comment by srkukarni** [comment]
run java8 tests

File: pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/ContextImpl.java (hunk 1 above)
Comment (srkukarni) [reply pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/ContextImpl.java]: Fixed

**Comment by srkukarni** [comment]
run integration tests
run cpp tests

**Comment by srkukarni** [comment]
run integration tests

**Comment by srkukarni** [comment]
run integration tests

**Comment by merlimat** [comment]
run java8 tests
run integration tests

**Comment by srkukarni** [comment]
run java8 tests
run integration tests

**Comment by srkukarni** [comment]
run integration tests
run java8 tests

**Comment by srkukarni** [comment]
run java8 tests

**Comment by srkukarni** [comment]
run java8 tests

**Comment by merlimat** [comment]
Merged in 2.3.1 at 
8aa63aa

**Pull request merged**